### ISSUE TITLE
Catboost fit_transform method is broken.

### ISSUE THREAD

**Issue by PraveshKoirala** [scikit-learn-contrib/category_encoders#351]
## TLDR
When called fit_transform, the output is shuffled and not in sync with the input. This does not occur when called transform() after fitting. This tidbit should either be explained in documentation or should be solved to give the expected (i.e. ordered) results (order being values in the same order as the categories).

## Example
```
cats = pd.DataFrame({'X':['a', 'a', 'd', 'd', 'd', 'f', 'f'], 'id': ['a', 'a', 'd', 'd', 'd', 'f', 'f'], 'y' : [0, 0, 1, 0, 1, 1, 1]})
encoder = ce.CatBoostEncoder(cols=['X'])
encoder.fit(cats, y, return_df=True)
encoder.transform(cats)
```
gives

![image](https://user-images.githubusercontent.com/2066100/165650004-d59c5726-1bdc-42b4-b3ae-d4b40721e0c8.png)

```
cats = pd.DataFrame({'X':['a', 'a', 'd', 'd', 'd', 'f', 'f'], 
                     'id': ['a', 'a', 'd', 'd', 'd', 'f', 'f'], 
                     'y' : [0, 0, 1, 0, 1, 1, 1]})
encoder = ce.CatBoostEncoder(cols=['X'])
encoder.fit_transform(cats, y, return_df=True)
```

gives (NOTICE DIFFERENT VALUES FOR TWO a's):
![image](https://user-images.githubusercontent.com/2066100/165649879-b8ef64a0-ecc1-4258-addb-50114a733d53.png)

**Comment by PaulWestenthanner** [comment]
Hi @PraveshKoirala 

this is not a bug.
`fit_transform` calls `transform(X, y)` with the target information. As stated in catboost transform documentation

 > y : array-like, shape = [n_samples] when transform by leave one out
 >           None, when transform without target information (such as transform test set)

This always leaves out the current value. Hence we expect to see some differences. Indeed 
``` python
cats = pd.DataFrame({'X':['a', 'a', 'd', 'd', 'd', 'f', 'f'], 'id': ['a', 'a', 'd', 'd', 'd', 'f', 'f'], 'y' : [0, 0, 1, 0, 1, 1, 1]})
encoder = ce.CatBoostEncoder(cols=['X'])
encoder.fit(cats, y, return_df=True)
encoder.transform(cats, y)
```
gives the same result as `fit_transform`.
Does this make sense for you?

**Comment by PaulWestenthanner** [comment]
I'm not quite sure though why our implementation uses this cumsum and cumcount. With this the output is dependent on the ordering of the input. I'm not super deep into catboost algorithm but I know that our implementation differs at some points from the catboost paper (and the "official" yandex implementation).
Feel free to dig into it if you have time. These should be the relevant lines:
https://github.com/scikit-learn-contrib/category_encoders/blob/12e20486f4422a56c802a0e04163a896271d4107/category_encoders/cat_boost.py#L269-L280

**Comment by glevv** [comment]
I think we should update CatBoost documentation with this and #337 taken into account
